examples/keras_io/vision/metric_learning.py

```diff
@@ -0,0 +1,323 @@
+        anchors, positives = data[0], data[1]
+
+        with tf.GradientTape() as tape:
+            # Run both anchors and positives through model.
+            anchor_embeddings = self(anchors, training=True)
+            positive_embeddings = self(positives, training=True)
+
+            # Calculate cosine similarity between anchors and positives. As they have
+            # been normalised this is just the pair wise dot products.
+            similarities = keras.ops.einsum(
+                "ae,pe->ap", anchor_embeddings, positive_embeddings
+            )
+
+            # Since we intend to use these as logits we scale them by a temperature.
+            # This value would normally be chosen as a hyper parameter.
+            temperature = 0.2
+            similarities /= temperature
+
+            # We use these similarities as logits for a softmax. The labels for
+            # this call are just the sequence [0, 1, 2, ..., num_classes] since we
+            # want the main diagonal values, which correspond to the anchor/positive
+            # pairs, to be high. This loss will move embeddings for the
+            # anchor/positive pairs together and move all other pairs apart.
+            sparse_labels = keras.ops.arange(num_classes)
+            loss = self.compiled_loss(sparse_labels, similarities)
+
+        # Calculate gradients and apply via optimizer.
+        gradients = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+
+        # Update and return metrics (specifically the one for the loss value).
+        self.compiled_metrics.update_state(sparse_labels, similarities)
+        return {m.name: m.result() for m in self.metrics}
+
+
+"""
+Next we describe the architecture that maps from an image to an embedding. This model
+simply consists of a sequence of 2d convolutions followed by global pooling with a final
+linear projection to an embedding space. As is common in metric learning we normalise the
+embeddings so that we can use simple dot products to measure similarity. For simplicity
+this model is intentionally small.
+"""
+
+inputs = layers.Input(shape=(height_width, height_width, 3))
+x = layers.Conv2D(filters=32, kernel_size=3, strides=2, activation="relu")(inputs)
+x = layers.Conv2D(filters=64, kernel_size=3, strides=2, activation="relu")(x)
+x = layers.Conv2D(filters=128, kernel_size=3, strides=2, activation="relu")(x)
+x = layers.GlobalAveragePooling2D()(x)
+embeddings = layers.Dense(units=8, activation=None)(x)
+embeddings = layers.UnitNormalization()(embeddings)
+
+model = EmbeddingModel(inputs, embeddings)
+
+"""
+Finally we run the training. On a Google Colab GPU instance this takes about a minute.
+"""
+
+model.compile(
+    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+)
+
+# history = model.fit(AnchorPositivePairs(num_batches=1000), epochs=20)
```

Comment:
remove the comment and fix the number of epochs = 20

Reply:
done